### PR TITLE
Adding preflight validation for search engine

### PR DESCRIPTION
**Pull request by vinay-satish**
Signed-off-by: Vinay Satish <vinay.satish@progress.com>

### Description

The search will not work as expected if the indexes are configured to read-only. This PR adds a preflight validation to make sure that indexes are not read-only

### Issues Resolved
https://github.com/chef/chef-server/issues/3109

Doc PR - https://github.com/chef/chef-server/pull/3182

### Check List

- [ ] New functionality includes tests
- [ ] All buildkite tests pass
- [ ] Full omnibus build and tests in buildkite pass
- [ ] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/main/CONTRIBUTING.md#developer-certification-of-origin-dco>
- [ ] PR title is a worthy inclusion in the CHANGELOG

### CODE REVIEW

**Comment by vinay-satish** [comment]
Adhoc build - https://buildkite.com/chef/chef-chef-server-main-omnibus-adhoc/builds/3938

**Comment by jashaik** [comment]
Pipielines -
Adhoc - https://buildkite.com/chef/chef-chef-server-main-omnibus-adhoc/builds/3967

File: omnibus/files/server-ctl-cookbooks/infra-server/libraries/preflight_indexing_validator.rb [hunk 1]
```diff
@@ -259,4 +275,17 @@ def err_INDEX007_bad_queue_mode
                 in #{CHEF_SERVER_CONFIG_FILE}
     EOM
   end
+
+  def err_INDEX008_bad_elasticsearch_config
+    <<~EOM
+
+      INDEX007: Invalid elasticsearch configuration
+
+      The elasticsearch index configuration is set to read only.
```
Comment (tas50) [inline omnibus/files/server-ctl-cookbooks/infra-server/libraries/preflight_indexing_validator.rb:284]:
```suggestion
      The Elasticsearch index configuration is set to read only.
```

File: omnibus/files/server-ctl-cookbooks/infra-server/libraries/preflight_indexing_validator.rb [hunk 2]
```diff
@@ -259,4 +275,17 @@ def err_INDEX007_bad_queue_mode
                 in #{CHEF_SERVER_CONFIG_FILE}
     EOM
   end
+
+  def err_INDEX008_bad_elasticsearch_config
+    <<~EOM
+
+      INDEX007: Invalid elasticsearch configuration
```
Comment (tas50) [inline omnibus/files/server-ctl-cookbooks/infra-server/libraries/preflight_indexing_validator.rb:282]:
```suggestion
      INDEX007: Invalid Elasticsearch configuration
```

**Comment by tas50** [comment]
https://buildkite.com/chef/chef-chef-server-main-omnibus-adhoc/builds/3980

**Comment by jashaik** [comment]
https://buildkite.com/chef/chef-chef-server-main-omnibus-adhoc/builds/3986

**Comment by jashaik** [comment]
https://buildkite.com/chef/chef-umbrella-main-chef-server/builds/1090

**Review by marcparadise — APPROVED** [review]

File: omnibus/files/server-ctl-cookbooks/infra-server/libraries/preflight_indexing_validator.rb [hunk 3]
```diff
@@ -130,6 +131,23 @@ def verify_erchef_config
     end
   end
 
+  def verify_search_engine_permissions
+    if node['private_chef']['opscode-erchef']['search_provider'] != 'solr'
+      search_provider_url = helper.solr_url
+      begin
+        client = Chef::HTTP.new("#{search_provider_url}/_all/_settings")
+        response = JSON.parse(client.get(''))
+        if response['chef']['settings']['index'].key?('blocks')
```
Comment (marcparadise) [inline omnibus/files/server-ctl-cookbooks/infra-server/libraries/preflight_indexing_validator.rb:140]: These can be combined with `response.dig('chef', 'settings', 'index', 'blocks', 'read_only_allow_delete)`

**Review by lbakerchef — CHANGES_REQUESTED** [review]

File: omnibus/files/server-ctl-cookbooks/infra-server/libraries/preflight_indexing_validator.rb [hunk 4]
```diff
@@ -130,6 +131,23 @@ def verify_erchef_config
...
+          if response['chef']['settings']['index']['blocks']['read_only_allow_delete']
+            fail_with err_INDEX008_bad_elasticsearch_config
+          end
+        end
+      rescue StandardError
+        Chef::Log.debug('Elasticsearch is not yet up so skipping search engine permissions check.')
```
[the remark this answers is not in the document]
Comment (jashaik) [reply omnibus/files/server-ctl-cookbooks/infra-server/libraries/preflight_indexing_validator.rb:146]: I will look into it. Thanks for finding it.

**Review by PrajaktaPurohit — APPROVED** [review]

File: omnibus/files/server-ctl-cookbooks/infra-server/libraries/preflight_indexing_validator.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (lbakerchef) [reply omnibus/files/server-ctl-cookbooks/infra-server/libraries/preflight_indexing_validator.rb]: I saw your screenshot triggering the error - thanks.  I ran an umbrella to confirm it does not trigger under normal circumstances.

https://buildkite.com/chef/chef-chef-server-main-omnibus-adhoc/builds/4010
https://buildkite.com/chef/chef-umbrella-main-chef-server/builds/1098

**Review by lbakerchef — APPROVED** [review]

**Pull request merged**